edx/edx-enterprise
# mgmt command to backfill enterprise role assignment data

https://openedx.atlassian.net/browse/ENT-4099

Adds a management command for updating enterprise admin and learner role assignments
with a foreign key to an EnterpriseCustomer, or an explicit boolean 
flag indicating that the assignment grants the role to the user
for every EnterpriseCustomer.

**Testing Instructions**
First, make sure to apply migrations for `enterprise` - we've added a config model that can define the arguments passed to this new management command.  This is generally how we parameterize management commands that run via Jenkins and typically need to accept arguments.

Do a dry run for admin assignments of some customer: 
```
./manage.py lms update_role_assignments_with_customers --role=enterprise_admin --dry-run --enterprise-customer-uuid=00000000-1111-2222-3333-444444444444
```
Do a real run for all operators: 
```
./manage.py lms update_role_assignments_with_customers --role=enterprise_openedx_operator
```
Process everything for everyone:
```
./manage.py lms update_role_assignments_with_customers
```
Undo the changes: 
```
./manage.py lms update_role_assignments_with_customers --undo
```

Define some arguments in the database now at http://localhost:18000/admin/enterprise/updateroleassignmentswithcustomersconfig/

Run the command using arguments from from that config model:
```
./manage.py lms update_role_assignments_with_customers --args-from-database
```

**Merge checklist:**
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.

**Post merge:**
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.
